### PR TITLE
DAO V3: Nouns Fork, propose by sigs, updatable period, objection period, snapshot votes at vote start

### PR DESCRIPTION
**Pull request by eladmallel**
## Change summary

1. **Proposal editing**: allowing proposers to update their proposal's transactions and text description, during the Updatable period only, which is the state upon proposal creation. Editing also works with signatures, assuming the proposer is able to accumulate signatures from the same signers.
2. **Propose by signature**: allowing Nouners and delegates to pool their voting power towards submitting a proposal, by submitting their signature, instead of the current approach where sponsors must delegate their votes to help a proposer achieve threshold.
3. **Objection-only Period**: a conditional voting period that gets activated upon a last-minute proposal swing from defeated to successful, affording against voters more reaction time. Only against votes are possible during the objection period.
4. **Votes Snapshot After Voting Delay**: moving votes snapshot up, to provide Nouners with reaction time per proposal, to get their votes ready (e.g. some might want to move their delegations around). Currently the vote snapshot block is the proposal creation block.
5. **Nouns Fork**: any token holder can signal to fork (exit) in response to a governance proposal. If a quorum of 20% of tokens signals to exit, the fork will succeed. For a deeper understanding please read [the tech spec](https://github.com/verbsteam/nouns-fork-spec/blob/main/spec.md) and [this post](https://mirror.xyz/0x10072dfc23771101dC042fD0014f263316a6E400/iN0FKOn_oYVBzlJkwPwK2mhzaeL8K2-W80u82F7fHj8).

For more detail on all features except Nouns Fork, [see this spec](https://hackmd.io/@el4d/nouns-dao-v3-spec).

# Upgrade Plan

The Nouns governor is currently at V2. This version under audit is V3, and the upgrade can only happen through a proposal, since only the treasury (executor) contract has the permission to set the governor proxy's implementation address.

The Nouns Fork feature introduced in this version requires the ability to send DAO funds to fork DAOs without going through a proposal process; this requirement means we need a new treasury, and since the current treasury is not upgradable, our plan is to migrate funds to the new treasury, and make the new treasury upgradeable to minimize the risk of another such migration in future.

Furthermore, for several reasons it's important to maintain the ability to execute proposals on the original treasury, and for that reason we've added proposal functions that would still use it on the new governor version, e.g. `proposeOnTimelockV1` and `executeOnTimelockV1`.

The deployment consists of the following scripts:

1. [`DeployDAOV3NewContractsMainnet`] Deploy all the new contracts.
2. [`ProposeDAOV3UpgradeMainnet`] Propose the upgrade to V3 and all the config changes this upgrade entails:
   1. Send ETH to treasury v2; sending a fixed amount that is lower than the entire balance, to account for proposals that might be executing on treasury V1 around the same time.
   2. Set governor proxy's implementation to V3.
   3. Set V3's fork parameters.
   4. Set the proposal ID at which vote snapshot blocks change from proposal creation block to voting period start block.
   5. Set the auction house owner to treasury V2.
   6. Send stETH to treasury v2, while keeping a buffer of 1K stETH in v1.
   7. Transfer ownership of NounsToken to treasury V2.
   8. Transfer ownership of NounsToken's descriptor to treasury V2.
   9. Transfer ownership of auction house's proxy admin to treasury V2.
   10. Set the governor's admin and treasury addresses, making treasury v2 the main executor contract and the governor's admin.
3. [`ProposeTimelockMigrationCleanupMainnet`] Propose additional txs to treasury V1. This proposal will be proposed once `ProposeDAOV3UpgradeMainnet` has been successfully executed. This gives time for proposals using TokenBuyer (asking for USDC)
   to execute successfully.
   This script includes the following:
   1. Transfer ETH leftover from the upgrade proposal to treasury V2.
   2. Approve treasury V2 to transfer LilNouns NFTs owned by treasury V1.
   3. Transfer TokenBuyer ownership to treasury v2.
   4. Transfer Payer ownership to treasury v2 (Payer is a helper contract part of the TokenBuyer).
   5. Approve a ERC20Transferer helper contract to transfer the DAO's stETH tokens.
   6. Call the transferer's function to move all remaining stETH tokens from treasury v1 to v2.
   7. Transfer all wstETH tokens from treasury V1 to V2.
   8. Transfer all rETH tokens from treasury V1 to V2.
4. [`ProposeENSReverseLookupConfigMainnet`] Propose post-upgrade on treasury V2 to set its ENS reverse lookup to `nouns.eth`.
   1. Worth noting this proposal should only be executed after the owner of `nouns.eth` updates the main record to point to treasury V2 instead of V1. The relevant person from the founders is ready to execute the ENS changes in time.

### CODE REVIEW

File: packages/nouns-contracts/contracts/governance/NounsDAOInterfaces.sol [hunk 1]
```diff
@@ -704,7 +704,7 @@ contract NounsDAOStorageV3 {
         uint256 creationBlock;
         address[] signers;
         uint256 updatePeriodEndBlock;
-        uint256 objectionPeriodEndBlock;
+        uint64 objectionPeriodEndBlock;
```
Comment (davidbrai) [inline packages/nouns-contracts/contracts/governance/NounsDAOInterfaces.sol:707]: @eladmallel why is this saving gas?

File: packages/nouns-contracts/contracts/governance/NounsDAOInterfaces.sol [hunk 2]
```diff
@@ -703,7 +703,7 @@ contract NounsDAOStorageV3 {
         /// @notice The block at which this proposal was created
         uint256 creationBlock;
         address[] signers;
-        uint256 updatePeriodEndBlock;
+        uint64 updatePeriodEndBlock;
```
Comment (davidbrai) [inline packages/nouns-contracts/contracts/governance/NounsDAOInterfaces.sol:706]: @eladmallel why is this saving gas?

File: packages/nouns-contracts/contracts/governance/fork/NounsDAOV3Fork.sol [hunk 3]
```diff
@@ -30,10 +30,16 @@ library NounsDAOV3Fork {
     error ERC20TransferFailed();
 
     /// @notice Emitted when someones adds nouns to the fork escrow
-    event EscrowedToFork(address indexed owner, uint256[] tokenIds, uint256[] proposalIds, string reason);
+    event EscrowedToFork(
+        address indexed owner,
+        uint256[] tokenIds,
+        uint256[] proposalIds,
+        string reason,
+        uint32 forkId
```
Comment (davidbrai) [inline packages/nouns-contracts/contracts/governance/fork/NounsDAOV3Fork.sol:38]: @eladmallel how about making `forkId` indexed?

File: packages/nouns-contracts/test/foundry/NounsDAOLogicV3/UpgradeToDAOV3ForkMainnetTest.t.sol [hunk 4]
```diff
@@ -238,6 +244,19 @@ contract UpgradeToDAOV3ForkMainnetTest is Test {
         assertEq(forkDao.quorumVotesBPS(), 1000);
     }
 
+    function test_timelockV1CleanupProposal() public {
+        assertGt(address(NOUNS_TIMELOCK_V1_MAINNET).balance, 2919 ether);
+
+        proposalId = new ProposeTimelockMigrationCleanupMainnet().run();
+        executeUpgradeProposal();
```
Comment (davidbrai) [inline packages/nouns-contracts/test/foundry/NounsDAOLogicV3/UpgradeToDAOV3ForkMainnetTest.t.sol:252]: @eladmallel nit: this was a bit confusing. rename to `voteAndExecuteProposal` ?

File: packages/nouns-contracts/test/foundry/NounsDAOLogicV3/UpgradeToDAOV3ForkMainnetTest.t.sol [hunk 5]
```diff
@@ -238,6 +244,19 @@ contract UpgradeToDAOV3ForkMainnetTest is Test {
...
+
+        assertEq(nouns.owner(), address(timelockV2));
+        assertEq(Ownable(DESCRIPTOR_MAINNET).owner(), address(timelockV2));
+        assertEq(Ownable(AUCTION_HOUSE_PROXY_ADMIN_MAINNET).owner(), address(timelockV2));
+        assertEq(address(NOUNS_TIMELOCK_V1_MAINNET).balance, 0);
```
Comment (davidbrai) [inline packages/nouns-contracts/test/foundry/NounsDAOLogicV3/UpgradeToDAOV3ForkMainnetTest.t.sol:257]: @eladmallel would be nice to test that `timelockV2` got that eth

**Comment by solimander** [comment]
Can `fork-audit-diffs` be dropped now?

**Comment by eladmallel** [comment]
> Can `fork-audit-diffs` be dropped now?

removing

**Review by solimander — APPROVED** [review]

**Pull request merged**